### ISSUE TITLE
Video issue: Momentarily freezes during video playback (HWA)

### ISSUE THREAD

**Issue by THEtomaso** [MoonchildProductions/Pale-Moon#982]
With MSE enabled, certain videos will freeze momentarily during playback.
Sound is unaffected though, and when the video unfreezes, it catches up with the sound immediately and gets back in sync.
Not everyone seems to be able to confirm this, so the issue seems to be system dependent somehow.

Examples:
http://www.vgtv.no/#!/video/137889
http://tv.aftonbladet.se/abtv/articles/210961/

As a workaround, setting Pale Moon's User Agent Mode to "Native" fixes it.
However, when using this workaround, the videos doesn't play in HD by default, so you'll have to change the resolution manually.

(Win8.1Pro,x64 / PM27.*,x64)

**Comment by wolfbeast** [comment]
I find such up-to-a-second pauses to be common for all news sites, regardless of MSE being in use or not.

**Comment by adisib** [comment]
I encounter this with YouTube often, but I seem to remember it always being a problem before Tycho as well, when MSE wasn't implemented. I find them to be more of at-least-a-second pauses, rather than up-to-a-second pauses for me though.
It generally seems to occur if I pause a video, watch a different video, and start the video again later. I generally just press the left arrow key to seek back a for a few seconds to force it to instantly sync, because it is faster than waiting for the sync and and I still get to see all of the video frames.

**Comment by THEtomaso** [comment]
If I disable MSE, they play just fine.
I've tested this many, many times.

**Comment by wolfbeast** [comment]
Please see if the current unstable channel improves this.

**Comment by THEtomaso** [comment]
Actually, I tried the latest unstable yesterday (v27.3.0a1,2017-04-09), and the problem persists.

**Comment by THEtomaso** [comment]
OK, I've done some further tests now, and I've found 3 workarounds that fixes this issue:

Workaround #1:
Setting Pale Moon's User Agent Mode to "Native".

Workaround #2:
Disable H/W acceleration in Pale Moon.

Workaround #3:
Configure the Nvidia driver so that colour adjustments are done by the player.
This is actually the default setting too, but unfortunately, as a result this will reduce the videos dynamic range conciderably, and all HTML5 videos will have this washed out look to them, like you can see from my screenshots here:
https://forum.palemoon.org/viewtopic.php?f=3&t=9828#p78281
https://forum.palemoon.org/viewtopic.php?f=3&t=9828#p78291
I've posted information on how to change the setting through Nvidia's control panel here: 
https://forum.palemoon.org/viewtopic.php?f=29&t=13785&p=98940#p98940
This was also recently decribed by Greywool here:
https://forum.palemoon.org/viewtopic.php?f=3&t=15333&p=111091#p111091
Other browsers (like SeaMonkey) does not have this problem..
Videos won't freeze when colour adjustments are done by the driver, and they will look just as good when colour adjustments are done by the player.

**Comment by THEtomaso** [comment]
Apparently, the issue with limited dynamic range in HTML5 videos was fixed in Firefox Nightly v49.0.0.5960:
https://bugzilla.mozilla.org/show_bug.cgi?id=1125844
https://bugzilla.mozilla.org/show_bug.cgi?id=1161349

**Comment by wolfbeast** [comment]
So, it's very specific to NV and letting the driver do post-processing... 
Workaround 1 likely falls back to a different player (maybe flash or non-MSE/different codec)

Workaround 2 and 3 would prevent GPU post-processing. Unless the color-correction change in the bugs mentioned actually avoid the problem in NV's driver, there isn't really much that can be done on the browser's side since it won't be in our code.

**Comment by wolfbeast** [comment]
Are you using d3d9 or d3d11 layers, @THEtomaso ? Did you set `layers.prefer-d3d9`?
Can you post the exact output of your troubleshooting information "Graphics" section as text?

**Comment by wolfbeast** [comment]
Also:

> Not everyone seems to be able to confirm this, so the issue seems to be system dependent somehow.

That is expected. HWA video playback is very dependent on hardware, and on some systems, other browsers will simply perform better than Pale Moon. It's an inevitability of this kind of task, because video is not just played back, it also has to be fit into rendered content that is, itself, also drawn hardware-accelerated on a surface.

**Comment by THEtomaso** [comment]
> Did you set layers.prefer-d3d9?

No.

--

> Can you post the exact output of your troubleshooting information "Graphics" section as text?

Adapter Description: NVIDIA GeForce GTX 780 Ti
Adapter Drivers: nvd3dumx,nvwgf2umx,nvwgf2umx nvd3dum,nvwgf2um,nvwgf2um
Adapter RAM: 3072
Device ID: 0x100a
Direct2D Enabled: true
DirectWrite Enabled: true (6.3.9600.18592)
Driver Date: 3-16-2017
Driver Version: 21.21.13.7892
GPU #2 Active: false
GPU Accelerated Windows: 1/1 Direct3D 11 (OMTC)
Subsys ID: 28813842
Vendor ID: 0x10de
WebGL Renderer: Google Inc. -- ANGLE (NVIDIA GeForce GTX 780 Ti Direct3D11 vs_5_0 ps_5_0)
windowLayerManagerRemote: true
AzureCanvasBackend: direct2d 1.1
AzureContentBackend: direct2d 1.1
AzureFallbackCanvasBackend: cairo
AzureSkiaAccelerated: 0

**Comment by THEtomaso** [comment]
Should I open a separate issue for Pale Moon's dynamic range issue with HTML5 videos?

**Comment by wolfbeast** [comment]
"Momentary freezes" != "dynamic range"; what do you think? ;-)

**Comment by THEtomaso** [comment]
> ¨"Momentary freezes" != "dynamic range"; what do you think?

I ment the issue that was fixed in Firefox two years ago, which causes HTML5 videos to be shown with limited dynamic range when colour adjustments are done by the player:
https://github.com/MoonchildProductions/Pale-Moon/issues/982#issuecomment-293199598
If this was fixed in Pale Moon, then I wouldn't have to use the Nvidia driver for colour adjustments anymore.
Or are you saying that using full dynamic range is what causes the freezes in the place, regardless of where the colour adjustments are done?

**Comment by THEtomaso** [comment]
Oh, btw; I forgot to include the most obvious workaround for the freezing issue, so I've edited my post here:
https://github.com/MoonchildProductions/Pale-Moon/issues/982#issuecomment-293193601
:)

**Comment by THEtomaso** [comment]
I've edited my post here:
https://github.com/MoonchildProductions/Pale-Moon/issues/982#issuecomment-293193601
..noting that disabling MSE also fixes the freezing issue, like I've pointed out before.

**Comment by THEtomaso** [comment]
Unfortunately, the MSE changes in Pale Moon v27.3.0a1 (2017-04-16) has made this issue even worse.
The momentarily freezes seems to happen twice as often now.

On the positive side, disabling MSE still fixes the issue.
..and unlike with previous builds, the videos doesn't play in SD resoultion by default anymore when MSE is disabled! :)

**Comment by THEtomaso** [comment]
I just gave v27.4.0a1 (2017-04-30) a try, playing the videos that I linked to in the first post..

The videos @ vgtv.no plays just fine now!

However, the videos @ aftonbladet.se still suffers from freezes.
Also, for these videos, configuring the Nvidia driver so that colour adjustments are done by the player, doesn't help at all, like it used to for the vgtv videos.

Many thanks to trav90, which is working hard with this stuff.
At least, something seems to have payed off in regard to the vgtv videos. :)
I'll continue to try the unstable builds, and let you know if anything changes for the aftonbladet videos too.
